### ISSUE TITLE
Memoization of functions is awkwardly sensisitive to irrelevant source code changes

### ISSUE THREAD

**Issue by benclifford** [Parsl/parsl#1591]
**Describe the bug**
Sometimes functions are passed as parameters to app functions, and so are used as material for the memoization hash (`task_record['hashsum']`). This does not work well, resulting in apparently-irrelevant changes to source code causing memoization to miss previously computed results.

This can happen both in when a user passes in a function (for example a wrapper function) explicitly, but more importantly happens when the raw app function is passed into one of the several internal parsl wrapper functions - so even if a workflow user does not pass in a function explicitly, function memoization happens.

The current master behaviour is to checkpoint functions using the default serialization-based mechanism, registered here:
```
@id_for_memo.register(types.FunctionType)
def id_for_memo_serialize(obj, output_ref=False):
```
which in turn uses `CannedFunction` from `parsl/executors/serialize/canning.py` which in turn uses the code pickling in `parsl/executors/serialize/codeutil.py`. 

`codeutil.py` includes lots of fields into this result:
```
    args = [co.co_argcount, co.co_nlocals, co.co_stacksize,
            co.co_flags, co.co_code, co.co_consts, co.co_names,
            co.co_varnames, co.co_filename, co.co_name, co.co_firstlineno,
            co.co_lnotab, co.co_freevars, co.co_cellvars]

```

so at the very least, the memoization for a function depends on the line number at which it is defined, and so adding a one line comment at the top of a source file invalidates the memoization hash of every function defined in that source file.

**Expected behavior**
This is up for debate.

My favoured option at the moment is that function memo hashing should be quite tolerant, maybe using just the function module/name, and not be dependent at all on how that function is defined. Dependencies on the code inside the function mean that changing a function definition directly causes memo invalidation, but changing any helper functions used by that function do not cause a memo invalidation. This is unintuitive.

I've modified the `lsst-dm-202002` branch to have a lighter form of function memo hashing that only looks at module and name, and have been using that with @tomglanzman

**Comment by benclifford** [comment]
PR #1766 switches function memoization to using only function name and module (usually, the python source file, but I'm unsure how that works in notebooks)

**Comment by yadudoc** [comment]
I agree that memoization should not be dependent on the line no of the function, or its docs, and various other notebook related variables. 

I'm not convinced that we should be ignoring the function body. We do not interrogate the code in depth to check if the functions being called have changed, so there are ways for the user to make changes without Parsl noticing. That's okay, there are limits to what we can interrogate. But a change made at the function body where we could easily say the function has changed should not be ignored and should invalidate the usage of previous results. I see two options here: 
1. Let the user mark functions to ignore function body
2. Check code and ignore comments/docs to make sure we only look at code changes. I don't know if this solves the issue.

**Comment by kylechard** [comment]
I like Ben's suggestion of the default model using only the app name.  I think this puts the decision on the user to manage whether or not they want to use checkpoints for a particular app.  We could offer an option to hash the body, but as described there are so many limitations there that it feels like something users need to understand fully before enabling.

**Comment by annawoodard** [comment]
Under @benclifford's model, what is the proposed way of dealing with situations where you have a large, partially processed workflow, and you need to make a change to an intermediate app? You'd like to re-run those results, but you don't want to disable caching (to force it to re-run), because after it is re-run the first time, you want it to be checkpointed. Do we expect users to change the name of their app every time they tweak the code?

**Comment by benclifford** [comment]
@annawoodard I've been deleting my checkpoint files.

There are also usecases where we've wanted to continue from a checkpoint with an app changed, without throwing away earlier work.

Targeted invalidation of specific checkpoints is something harder (if you want to delete results for just one particular app, for example - because that info isn't stored in the checkpoint DB in unhashed form) and is a problem if you change your app by (for example) upgrading executables in any of the implementations we've talked about.

My main goal is that I want to user-facing rules for when checkpoints are invalidated to be simple to understand and reason about.

**Comment by danielskatz** [comment]
cc @charmoniumQ
